### ISSUE TITLE
Discuss Azure Event Hub SDK 5.0

### ISSUE THREAD

**Issue by tkoster** [noopkat/electric-io#196]
Hi again Suz and the noopcrew! :wave:

I thought I would create this issue to discuss the breaking changes caused by the upgrade to `@azure/event-hubs@5` proposed by renovate in #192 , as I did some investigation into this that I'd like to share.

Feel free to elaborate and make corrections :sweat_smile:

## Background

The :zap: electric-io server uses two libraries to access the IoT Hub:

1. the [Event Hub SDK](https://www.npmjs.com/package/@azure/event-hubs) is used to subscribe to device-to-cloud messages to show on the dashboard
2. the [IoT Hub SDK](https://www.npmjs.com/package/azure-iothub) is used to list devices and send cloud-to-device messages

This issue is about the first one. At the time of writing, electric-io uses `@azure/event-hubs@2.1.1`. The latest version is `@azure/event-hubs@5.0.1`.

Of the many changes, two seem to affect electric-io adversely: the splitting of `EventHubClient` into `EventHubConsumerClient` and `EventHubProducerClient`, and the discontinuation of support for IoT Hub connection strings when creating Event Hub clients.

## Splitting client classes

The splitting of the client classes should hopefully require only fairly mechanical changes in the electric-io server. Most importantly, this should present no breaking change to electric-io users.

## Connection strings

Both IoT Hub and Event Hub use connection strings, but they are incompatible.

However, older versions of the Event Hub SDK `@azure/event-hubs` include a very convenient function, `EventHubClient.createFromIotHubConnectionString` that allowed you to use an IoT Hub connection string to connect to the Event Hub compatible endpoint in the IoT Hub transparently. Electric-io takes advantage of this to simplify the user experience: users only need one connection string - the IoT Hub connection string.

See [lib/liveHub.js](https://github.com/noopkat/electric-io/blob/554636cb806ced0f7ec31256fef86485e0c2eec1/lib/liveHub.js#L25).

Unfortunately, this function was removed from the SDK in late 2019, in https://github.com/Azure/azure-sdk-for-js/pull/5311.

The [documentation for the SDK](https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-event-hubs/5.0.0/index.html#use-eventhubconsumerclient-to-work-with-iothub) now recommends using the Event Hub-compatible endpoint connection string directly. For electric-io, I think this would mean the user would need to supply two connection strings, as it can no longer rely on the SDK to get one from the other. This would be an unfortunate breaking change for electric-io users.

**Comment by tkoster** [comment]
### Reimplementing the previous connection string behaviour in electric-io

I did some research into how the original `createFromIotHubConnectionString` function worked, with the aim to perhaps reimplement it in electric-io so that we can manage the deprecation on our own terms and allow users to continue with just the one IoT Hub connection string.

This turned out to be a very deep rabbit hole (this cliche also has foreshadowing :nerd_face:).

After some source diving, basically it looks like `createFromIotHubConnectionString` uses the IoT Hub connection string to access a management interface on the IoT Hub, which it uses to get the details of the Event Hub-compatible endpoint. The client then constructs an Event Hub connection string from that and connects to the Event Hub.

#### Network capture

The library is complex, so instead I planned to reverse engineer what `createFromIotHubConnectionString` does from network traffic and see if I could reproduce the basics.

I assumed that the management API in the IoT Hub was accessed over HTTPS. After hunting around looking for the requests, I realised that the only TLS connection in my capture from my electric-io server process did not contain any HTTP, but an AMQP connection instead :rabbit2: :scream:. It seems the Event Hub client is sending an RPC over AMQP to that management interface in the IoT Hub.

This message seems to be a reply that returns a SAS token:

![wireshark-iothub-amqp-sastoken](https://user-images.githubusercontent.com/9063335/75986881-fae4cb80-5f42-11ea-9412-a678e03099ea.png)

Here is one that seems to use a redirect to reveal the address of the Event Hub-compatible endpoint:

![wireshark-iothub-amqp-endpoint](https://user-images.githubusercontent.com/9063335/75987563-3764f700-5f44-11ea-98ce-7a45240564f6.png)

This agrees with what I found in the code that `@azure/event-hubs@2.1.1` used to do this: https://github.com/Azure/azure-sdk-for-js/blob/e2e2e71252f666d1c7bf80bdd65c532d710cee13/sdk/eventhub/event-hubs/src/iothub/iothubClient.ts

#### Conclusion

I believe the Event Hub client is using all these pieces to construct the Event Hub connection string.

Long story short, this is sooooo complicated that I don't think it's feasible (or wise) to reproduce this ourselves in electric-io :smile:. Sorry about the anticlimax :sweat_smile:

**Comment by tkoster** [comment]
I looks like an alternative way to get the Event Hub-compatible endpoint is through Azure Resource Manager and [@azure/arm-iothub](https://www.npmjs.com/package/@azure/arm-iothub/v/3.0.0). This is probably how the Azure Portal is able to display this. However, this requires an Azure credential, e.g. an interactive Azure Active Directory login into Azure or a service principal. I'm not a fan of this for the "getting started" experience.

**Comment by tkoster** [comment]
Hi @noopkat  :coffee: :v:

I know I said this earlier:

> this is sooooo complicated that I don't think it's feasible (or wise) to reproduce this ourselves in electric-io 😄

...but your heartbreak at the inconvenience that #197 causes users motivated me to look closer at reproducing `EventHubClient.createFromIotHubConnectionString`.

It might not be as complicated as I thought, so an alternative PR to #197 is on its way :gift: . Have a lovely day :smile:

**Comment by noopkat** [comment]
closing as @tkoster fixed this ✨ 🌸
